File: 121-best-time-to-buy-and-sell-stock/best-time-to-buy-and-sell-stock.py
class Solution(object):
    def maxProfit(self, prices):
        """
        :type prices: List[int]
        :rtype: int
        """
        maxFuture = [0] * len(prices)
        most = 0
        maxProfit = 0
        for i in range(len(prices)):
            if prices[len(prices) - i - 1] > most:
                most = prices[len(prices) - i - 1] 
            maxFuture[len(maxFuture) - i - 1] = most
        
        for i in range(len(prices)):
            if maxFuture[i] - prices[i] > maxProfit:
                maxProfit = maxFuture[i] - prices[i]

        return maxProfit

        

File: 121-best-time-to-buy-and-sell-stock/best-time-to-buy-and-sell-stock.cpp
class Solution {
public:
    int maxProfit(vector<int>& prices) {
        int maxSeen = prices[prices.size() - 1];
        int max = 0;

        for (int i = prices.size() - 1; i >= 0; i--){
            if (maxSeen - prices[i] > max){
                max = maxSeen - prices[i];
            }
            if (prices[i] > maxSeen){
                maxSeen = prices[i];
            }
        }
        return max;
    }
};